config/webpack.config.js

```diff
@@ -242,46 +144,175 @@ module.exports = env => {
+  var config = commonConfig();
+  var manifestFilename = 'manifest.json';
+  var bundleEntry = path.join(
+    __dirname,
+    '..',
+    'webpack/assets/javascripts/bundle.js'
+  );
+  config.context = path.resolve(__dirname, '..');
+  config.entry = {
+    bundle: { import: bundleEntry, dependOn: 'vendor' },
+    vendor: vendorEntry,
+  };
+  config.output = {
+    path: path.join(__dirname, '..', 'public', 'webpack'),
+    publicPath: '/webpack/',
+  };
+  var plugins = config.plugins;
+
+  // since CardTemplate uses CardExpansionContext between plugins, we need to make sure there is only one source of it, otherwise multiple contexts will be created and only one will be initialized correctly
+  const coreShared = {
+    './webpack/assets/javascripts/react_app/components/HostDetails/Templates/CardItem/CardTemplate/index.js': {
+      singleton: true,
+      eager: true,
+      shareKey:
+        'foreman/webpack/assets/javascripts/react_app/components/HostDetails/Templates/CardItem/CardTemplate/index.js',
+    },
+  };
+
+  plugins.push(
+    new ModuleFederationPlugin({
+      name: 'foremanReact',
+      shared: {
+        ...coreShared,
+      },
+    })
+  );
+  plugins.push(
+    new StatsWriterPlugin({
+      filename: manifestFilename,
     })
   );
+  config.plugins = plugins;
+  var rules = config.module.rules;
+  rules.push({
+    test: /\.(sa|sc|c)ss$/,
+    use: ['style-loader', 'css-loader', 'sass-loader'],
+  });
+  config.module.rules = rules;
+  return config;
+};
 
-  if (production) {
-    config.plugins.push(
-      new webpack.NoEmitOnErrorsPlugin(),
-      new UglifyJsPlugin({
-        uglifyOptions: {
-          compress: { warnings: false },
-        },
-        sourceMap: true,
-      }),
-      new SimpleNamedModulesPlugin(),
-      new webpack.optimize.ModuleConcatenationPlugin(),
-      new webpack.optimize.OccurrenceOrderPlugin(),
-      new CompressionPlugin()
-    );
-    config.devtool = 'source-map';
-  } else {
-    config.plugins.push(
-      new webpack.HotModuleReplacementPlugin() // Enable HMR
+const pluginConfig = function(plugin) {
+  const pluginRoot = plugin.root;
+  const pluginName = plugin.name.replace('-', '_'); // module federation doesnt like -
+  var config = commonConfig();
+  config.context = path.join(pluginRoot, 'webpack');
+  config.entry = {};
+  var pluginEntries = {
+    './index': path.resolve(pluginRoot, 'webpack', 'index'),
+  };
+  plugin.entries.filter(Boolean).forEach(entry => {
+    pluginEntries[`./${entry}_index`] = path.resolve(
+      pluginRoot,
+      'webpack',
+      `${entry}_index`
     );
+  });
 
-    config.devServer = {
-      host: devServer.host,
-      port: devServer.port,
-      headers: { 'Access-Control-Allow-Origin': '*' },
-      hot: true,
-      stats: (process.env.WEBPACK_STATS || 'minimal'),
-    };
-    // Source maps
-    config.devtool = 'inline-source-map';
-  }
+  config.output = {
+    path: path.join(pluginRoot, 'public', 'webpack', pluginName),
```

Comment:
using pluginRoot here made RPM builds work, but I guess it breaks devel setups?

Reply:
seems this reliably hangs integration tests on GHA, nice

Reply:
We'll need to also make it so it will be served as I get `No route matches [GET] "/webpack/foreman_ansible/foreman_ansible_remoteEntry.js"` 
and add a git ignore to plugins :sweat_smile:

Reply:
huh, seems the CI failure is unrelated to my changes, as it also fails on plain `develop`: https://github.com/theforeman/foreman/actions/runs/7213337369/job/19653057456?pr=9955 :exploding_head:

Reply:
the issue is foreman-js 12.2.1 :see_no_evil:

Reply:
> We'll need to also make it so it will be served

This is fixed :)